src/nmea.cpp

```diff
@@ -0,0 +1,1015 @@
+
+		if (bufptr && *(++bufptr) != ',') { M = *(bufptr++); }
+
+		if (bufptr && *(++bufptr) != ',') { ground_speed = strtof(bufptr, &endp); bufptr = endp; }
+
+		if (bufptr && *(++bufptr) != ',') { N = *(bufptr++); }
+
+		if (bufptr && *(++bufptr) != ',') { ground_speed_K = strtof(bufptr, &endp); bufptr = endp; }
+
+		if (bufptr && *(++bufptr) != ',') { K = *(bufptr++); }
+
+		float track_rad = track_true * M_PI_F / 180.0f; // rad in range [0, 2pi]
+
+		if (track_rad > M_PI_F) {
+			track_rad -= 2.f * M_PI_F; // rad in range [-pi, pi]
+		}
+
+		float velocity_ms = ground_speed / 1.9438445f;
+		float velocity_north = velocity_ms * cosf(track_rad);
+		float velocity_east  = velocity_ms * sinf(track_rad);
+
+		_gps_position->vel_m_s = velocity_ms;
+		_gps_position->vel_n_m_s = velocity_north;
+		_gps_position->vel_e_m_s = velocity_east;
+		_gps_position->cog_rad = track_rad;
+		_gps_position->vel_ned_valid = true; /** Flag to indicate if NED speed is valid */
+		_gps_position->c_variance_rad = 0.1f;
+		_gps_position->s_variance_m_s = 0;
+
+		if (!_VEL_received) {
+			_VEL_received = true;
+		}
+	}
+
+	if (_sat_num_gga > 0) {
+		_gps_position->satellites_used = _sat_num_gga;
+
+	} else if (_SVNUM_received && _SVINFO_received && _FIX_received) {
+
+		_sat_num_gsv = _sat_num_gpgsv + _sat_num_glgsv + _sat_num_gagsv
+			       + _sat_num_gbgsv + _sat_num_bdgsv;
+		_gps_position->satellites_used = MAX(_sat_num_gns, _sat_num_gsv);
+	}
+
+	if (_VEL_received && _POS_received) {
+		ret = 1;
+		_gps_position->timestamp_time_relative = (int32_t)(_last_timestamp_time - _gps_position->timestamp);
+		_clock_set = false;
+		_VEL_received = false;
+		_POS_received = false;
+		_rate_count_vel++;
+		_rate_count_lat_lon++;
+	}
+
+	return ret;
+}
+
+int GPSDriverNMEA::receive(unsigned timeout)
+{
+	uint8_t buf[GPS_READ_BUFFER_SIZE];
+
+	/* timeout additional to poll */
+	gps_abstime time_started = gps_absolute_time();
+
+	int handled = 0;
+
+	while (true) {
+		int ret = read(buf, sizeof(buf), timeout);
+
+		if (ret < 0) {
+			/* something went wrong when polling or reading */
+			GPS_WARN("poll_or_read err");
+			return -1;
+
+		} else if (ret != 0) {
+
+			/* pass received bytes to the packet decoder */
+			for (int i = 0; i < ret; i++) {
+				int l = parseChar(buf[i]);
+
+				if (l > 0) {
+					handled |= handleMessage(l);
+				}
+			}
+
+			if (handled > 0) {
+				return handled;
+			}
+		}
+
+		/* abort after timeout if no useful packets received */
+		if (time_started + timeout * 1000 < gps_absolute_time()) {
+			return -1;
+		}
+	}
+}
+
+#define HEXDIGIT_CHAR(d) ((char)((d) + (((d) < 0xA) ? '0' : 'A'-0xA)))
+
+int GPSDriverNMEA::parseChar(uint8_t b)
+{
+	int iRet = 0;
+
+	switch (_decode_state) {
+	/* First, look for sync1 */
+	case NMEADecodeState::uninit:
+		if (b == '$') {
+			_decode_state = NMEADecodeState::got_sync1;
+			_rx_buffer_bytes = 0;
+			_rx_buffer[_rx_buffer_bytes++] = b;
+		}
```

Comment:
```
} else if (b == RTCM3_PREAMBLE && _rtcm_parsing) {
	_decode_state = NMEADecodeState::decode_rtcm3;
	_rtcm_parsing->addByte(b);

}
```

@bkueng, probably `_rtcm_parsing` should be added same as in ashtech driver for PPK (0acaf6185245382389f7f6019cee6537d8f5a8e6) .

Reply:
Yes, can you send a PR?

Reply:
Done: https://github.com/PX4/PX4-GPSDrivers/pull/88 (but haven't tested it)